include/runtime/instance/memory.h

```diff
@@ -31,6 +31,35 @@ namespace WasmEdge {
 namespace Runtime {
 namespace Instance {
 
+template <uint32_t const Length>
+void effective_memcpy(void *dest, const void *src) {
```

Comment:
If we use `volatile` keyword for the `reinterpret_cast`, can we possibly ignore the alignment check? In the CI, all unit tests succeed in debug mode but some fail with seg fault in release mode. Probably in release mode some optimizations are performed provided the assumption that all memory address are aligned. If we mark the cast as `volatile`, can we fix it?

Reply:
I have added an alignment check in the `effective_memcpy`. When the `src` and `dest` is not properly aligned, it will use memcpy instead. 

Now the previous failed CI can all get passed. I also ran the rustpython benchmark, and the performance is basically the same as the one without the alignment check, which means adding the alignment check can still gain almost the same performance improvement. Therefore I think we can still use the `effective_memcpy` instead.

Reply:
In this PR, we only have two main changes, refactor the code with const template and introduce effective_memcpy. And I just ran my benchmark again under different combination on my macbook with M1 chip.
| time | run1 | run2 | run3 | avg |
| --- | --- | --- | --- |  --- |
|Current | 118.08s | 117.88s | 118.87s | 118.28s |
|Current with effective_memcpy| 116.11s | 115.41s | 115.69s | 115.74s|
| const template with std::memcpy | 111.03s | 111.15s | 111.38s | 111.19s |
| const template with effective_memcpy | 110.87s | 111.10s | 112.04s | 111.34s| 

Seems that the performance is the basically the same for std::memcpy and effective_memcpy when we refactor with const template, while in the original code, introducing effective_memcpy does improve the performance. 

Maybe the compiler will do some optimization on std::memcpy when it knows the `Length` in advance in compile time?

Reply:
My guess is correct. 

Just try compile the code on `compiler explorer`. If we know the length of `memcpy` in advance, the compiler will not call `memcpy` and instead generate several load instruction. If the length is provided at runtime passed from the function parameter, the generated assembly code will call `memcpy`.

https://godbolt.org/z/5z1z5WPY7